sockeye/inference.py

```diff
@@ -322,18 +355,51 @@ def get_max_output_length_function(models: List[InferenceModel], num_stds: int)
     :param models: List of models.
     :param num_stds: Number of standard deviations to add as a safety margin. If -1, returned maximum output lengths
                      will always be 2 * input_length.
-    :return: Callable.
+    :param max_input_len: An optional overwrite of the maximum input length.
+    :return: The maximum input length and a function to get the output length given the input length.
     """
     max_mean = max(model.length_ratio_mean for model in models)
     max_std = max(model.length_ratio_std for model in models)
 
-    def get_max_output_length(input_length: int):
-        if num_stds < 0:
-            return input_length * C.TARGET_MAX_LENGTH_FACTOR
+    if num_stds < 0:
+        factor = C.TARGET_MAX_LENGTH_FACTOR
+    else:
         factor = max_mean + (max_std * num_stds)
+
+    supported_max_seq_len_source = min((model.max_supported_seq_len_source for model in models
+                                        if model.max_supported_seq_len_source is not None),
+                                       default=None)
+    supported_max_seq_len_target = min((model.max_supported_seq_len_target for model in models
+                                        if model.max_supported_seq_len_target is not None),
+                                       default=None)
+
+    training_max_seq_len_source = min(model.training_max_seq_len_source for model in models)
+
+    if max_input_len is None:
+        # Make sure that if there is a hard constraint on the maximum source or target length we never exceed this
+        # constraint. This is for example the case for learned positional embeddings, which are only defined for the
+        # maximum source and target sequence length observed during training.
+        if supported_max_seq_len_source is not None and supported_max_seq_len_target is None:
+            max_input_len = supported_max_seq_len_source
+        elif supported_max_seq_len_source is None and supported_max_seq_len_target is not None:
+            if np.ceil(factor * training_max_seq_len_source) > supported_max_seq_len_target:
+                max_input_len = int(np.floor(supported_max_seq_len_target / factor))
+            else:
+                max_input_len = training_max_seq_len_source
+        elif supported_max_seq_len_source is not None or supported_max_seq_len_target is not None:
+            if np.ceil(factor * supported_max_seq_len_source) > supported_max_seq_len_target:
+                 max_input_len = int(np.floor(supported_max_seq_len_target / factor))
```

Comment:
Minor: inconsistent indentation.

Reply:
what's inconsistent here?

Reply:
The line has one additional whitespace (compare the position of the m in the two max_input_len lines).

Reply:
huh, wow, great eyes!